create_layout.py

```diff
@@ -88,51 +88,125 @@
+
+  return (unchanged_artifacts, modified_artifacts, added_artifacts,
+      removed_artifacts)
+
+def create_material_rules(previous_link, current_link):
+  """Create generic material rules
+
+  - MATCH available materials with products from previous step (links must be
+      an ordered list) and
+  - ALLOW available materials if it is the first step in the list
+  - DELETE removed materials
+
+  Args:
+    previous_link: a link of previous step, including previous step's materials
+        and products
+    current link: a link of current step, including current step's materials
+        and products
+
+  Returns:
+    a list of material rules
+  """
+
+  expected_materials_rules = []
+  unchanged_artifacts, modified_artifacts, _, deleted_artifacts = \
+      changes_between_snapshots(current_link.materials, current_link.products)
+  curr_step_materials = unchanged_artifacts.union(modified_artifacts,
+      deleted_artifacts)
+  if previous_link is None:
+    # If current step is the initial step, we don't need to add MATCH rules.
+    # Instead, ALLOW all the existing artifacts except deleted artifact
+    for artifact in sorted(curr_step_materials.difference(deleted_artifacts)):
+      expected_materials_rules.append(["ALLOW", artifact])
 
-  * MATCH available materials with products from previous step (links must be an
-  ordered list) and
-  * ALLOW available materials if it is the first step in the
-  list
-  Returns a list of material rules
-  NOTE: Read header docstring for ideas for more complexity.  """
+  else:
+    # MATCH only those that already were in the previous step
+    for artifact in sorted(curr_step_materials.intersection(
+        previous_link.products)):
+      expected_materials_rules.append(
+          ["MATCH", artifact, "WITH", "PRODUCTS", "FROM", previous_link.name])
 
-  expected_materials = []
+    # Allow the rest by name
+    for artifact in sorted(curr_step_materials.difference(
+        previous_link.products).difference(deleted_artifacts)):
+      expected_materials_rules.append(["ALLOW", artifact])
 
-  if index == 0:
-    for material_name in links[index].materials.keys():
-      expected_materials.append(["ALLOW", material_name])
-    expected_materials.append(["DISALLOW", "*"])
+    # If MATCH rule and DELETE rule are applied to the same artifact
+    if deleted_artifacts.intersection(previous_link.products):
+      warnings.warn("DELET rule is moot because of the previous MATCH rule."
+          " Only the first rule for a given artifact has an effect")
 
-  else:
-    expected_materials = [
-        ["MATCH", "*", "WITH", "PRODUCTS", "FROM", links[index - 1].name]]
+  # DELETE deleted artifacts in the current step
+  for artifact in sorted(deleted_artifacts):
+    expected_materials_rules.append(["DELETE", artifact])
 
-  return expected_materials
+  # DISALLOW all other artifacts
+  expected_materials_rules.append(["DISALLOW", "*"])
 
+  return expected_materials_rules
```

Comment:
I think you could simplify the control flow in this function a little bit, here is what I would do:

```
- if there was a previous step, add match rules for all materials that were products in the previous step
- add delete rules for all deleted artifacts, 
  but warn for any delete rule that has no effect because of a previous match rule
- add allow rules for all remaining materials
- add disallow *
```